lib/ruled/client.lua

```diff
@@ -0,0 +1,717 @@
+    -- always override the user specified position with the default rule.
+    if props.placement then
+        -- It may be a function, so this one doesn't execute it like others
+        module.extra_properties.placement(c, props.placement, props)
+    end
+
+    -- Handle the geometry (since tags and screen are set).
+    if props.height or props.width or props.x or props.y or props.geometry then
+        module.extra_properties.geometry(c, nil, props)
+    end
+
+    -- Apply the remaining properties (after known race conditions are handled).
+    for property, value in pairs(props) do
+        if property ~= "focus" and property ~= "shape" and type(value) == "function" then
+            value = value(c, props)
+        end
+
+        local ignore = module.high_priority_properties[property] or
+            module.delayed_properties[property] or force_ignore[property]
+
+        if not ignore then
+            if module.extra_properties[property] then
+                module.extra_properties[property](c, value, props)
+            elseif type(c[property]) == "function" then
+                c[property](c, value)
+            else
+                c[property] = value
+            end
+        end
+    end
+
+    -- Apply all callbacks.
+    if callbacks then
+        for _, callback in pairs(callbacks) do
+            protected_call(callback, c)
+        end
+    end
+
+    -- Apply the delayed properties
+    for prop, handler in pairs(module.delayed_properties) do
+        if not force_ignore[prop] then
+            local value = props[prop]
+            if value ~= nil then
+                if type(value) == "function" then
+                    value = value(c, props)
+                end
+                handler(c, value, props)
+            end
+        end
+    end
+
+    -- Do this at last so we do not erase things done by the focus signal.
+    if props.focus and (type(props.focus) ~= "function" or props.focus(c)) then
+        c:emit_signal('request::activate', "rules", {raise=not capi.awesome.startup})
+    end
+end
+
+function module.execute(...) crules:_execute(...) end
+
+-- TODO v5 deprecate this
+function module.completed_with_payload_callback(c, props, callbacks)
+    module.execute(c, props, callbacks)
+end
+
+gobject._setup_class_signals(module)
+
+capi.client.connect_signal("request::manage", module.apply)
+
+-- Request rules to be added **after** all modules are loaded, but before the
+-- clients are managed. This allows module to listen to rules being added and
+-- either modify them or add their own in the right order.
+local function request_rules()
+    module.emit_signal("request::rules")
+end
+
+capi.client.connect_signal("scanning", request_rules)
+
+--@DOC_rule_COMMON@
+
+return setmetatable(module, {
+    __newindex = function(_, k, v)
+        if k == "rules" then
+            gdebug.deprecate(
+                "Use ruled.client.append_rules instead awful.rules.rules",
```

Comment:
s/instead/instead of/ ?  (but it is not clear to me currently if that would be correct then really)